bindings/rust/extended/s2n-tls/src/connection.rs

```diff
@@ -1408,27 +1412,41 @@ impl Connection {
         Ok(())
     }
 
-    /// Associates an arbitrary application context with the Connection to be later retrieved via
+    /// Associates arbitrary application contexts with the Connection to be later retrieved via
     /// the [`Self::application_context()`] and [`Self::application_context_mut()`] APIs.
     ///
-    /// This API will override an existing application context set on the Connection.
+    /// This API will add additional application context on top of existing contexts set on the Connection.
     ///
     /// Corresponds to [s2n_connection_set_ctx].
     pub fn set_application_context<T: Send + Sync + 'static>(&mut self, app_context: T) {
-        self.context_mut().app_context = Some(Box::new(app_context));
+        let context_type_id = TypeId::of::<T>();
+        self.context_mut()
+            .app_context
+            .insert(context_type_id, Box::new(app_context));
+    }
+
+    /// Remove a application context set on the Connection.
+    pub fn remove_application_context(
+        &mut self,
+        context_type_id: TypeId,
+    ) -> Option<Box<dyn Any + Send + Sync>> {
+        self.context_mut().app_context.remove(&context_type_id)
     }
 
     /// Retrieves a reference to the application context associated with the Connection.
     ///
-    /// If an application context hasn't already been set on the Connection, or if the set
-    /// application context isn't of type T, None will be returned.
+    /// If application context hasn't already been set on the Connection, or if the set
+    /// application context doesn't match the type id of type T, None will be returned.
     ///
     /// To set a context on the connection, use [`Self::set_application_context()`]. To retrieve a
     /// mutable reference to the context, use [`Self::application_context_mut()`].
     ///
     /// Corresponds to [s2n_connection_get_ctx].
-    pub fn application_context<T: Send + Sync + 'static>(&self) -> Option<&T> {
```

Comment:
I don't think we can change the parameters of an API, that's a breaking change? Hmmm I don't think we thought this API through fully. Probably we will need to discuss this.